File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/validation/validator/UniqueCollectTagValidator.java
package com.windea.demo.cloudcollect.core.validation.validator;

import com.windea.demo.cloudcollect.core.domain.entity.CollectTag;
import com.windea.demo.cloudcollect.core.service.CollectTagService;
import com.windea.demo.cloudcollect.core.validation.annotation.UniqueCollectTag;
import org.springframework.beans.factory.annotation.Autowired;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

/**
 * 唯一收藏标签的校验器。
 */
public class UniqueCollectTagValidator implements ConstraintValidator<UniqueCollectTag, CollectTag> {
	@Autowired
	private CollectTagService service;

	@Override
	public boolean isValid(CollectTag value, ConstraintValidatorContext context) {
		return !service.exists(value);
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/validation/validator/UniqueCollectTagValidator.kt
package com.windea.demo.cloudcollect.core.validation.validator

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.repository.*
import com.windea.demo.cloudcollect.core.validation.annotation.*
import org.springframework.beans.factory.annotation.*
import javax.validation.*

/**唯一收藏标签的校验器。*/
open class UniqueCollectTagValidator : ConstraintValidator<UniqueCollectTag, CollectTag> {
	@Autowired private lateinit var tagRepository: CollectTagRepository
	
	override fun isValid(value: CollectTag, context: ConstraintValidatorContext): Boolean {
		val name = value.name
		val userId = value.user.id ?: return false
		return !tagRepository.existsByNameAndUserId(name, userId)
	}
}
